### ISSUE TITLE
New diagnostic check for instance variables that are not initialized in constructor

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#1924]
Instance variables that are assigned values outside of the `__init__` method but not initialized in the `__init__` method can be a source of bugs. We may want to provide a new diagnostic check (e.g. `reportUninitializedInstanceVariable`) that detects and flags this case.

See recent request for a similar feature in mypy: https://github.com/python/mypy/issues/10552

**Comment by erictraut** [comment]
I implemented this feature, but I'm finding that it's so noisy (way too many false positives) that it has little utility. The biggest problem is with unit test code, which typically initializes its instance variables in a `startUp` method or something similar.

We'll discuss this internally, but I'm not sure it's worth adding this to the type checker.

**Comment by Rapptz** [comment]
I'm not sure if you're willing to hear outside opinions on this but for me it's very common to reuse code in `__init__` with a helper function like `from_data` that does JSON parsing from a `dict`. e.g.

```py
class A:
    def __init__(self, data: SomeTypedDict):
        self.from_data(data)
    def from_data(self, data: SomeTypedDict):
        self.foo: int = data['foo']
        self.bar: str = data['bar']
```

I use this pattern a lot since it allows me to remove duplication while also allowing the same code to work both as updating the instance from fresh network data and initialisation. If this were to lead to noisy diagnostics I'd just turn it off since it's not helpful.

**Comment by erictraut** [comment]
Thanks for that feedback. Yes, that's consistent with my thinking.

For now, I think we will not include this feature.

**Comment by Tomaz-Vieira** [comment]
This might be related; Pyright also doesn't pick up on use-before-assignment instance variables even if they are assigned inside the `__init__` body:

```python3
# pyright: strict

class Parent:
    def __init__(self, a: int):
        self.a = a

class Child(Parent):
    def __init__(self, b: int):
        super().__init__(a=self.b) # use before assignment, no errors
        self.b = b
```

**Comment by erictraut** [comment]
This is intended. Pyright does not (and really cannot) know about the order in which methods or functions are called. It will do local code flow analysis to detect unbound symbols, but it cannot determine this across functions.

**Comment by erictraut** [comment]
This fix is included in pyright 1.1.57, which I just published. It will also be included in the next release of pylance.
